processing/src/main/java/io/druid/segment/StringDimensionQueryHelper.java

```diff
@@ -0,0 +1,443 @@
+          ((DimensionSelector) dimInfo.selector).lookupName(id)
+      );
+    }
+  }
+
+  @Override
+  public Aggregator[][] getDimExtractionRowSelector(TopNParams params, TopNQuery query, Capabilities capabilities)
+  {
+    final BaseTopNAlgorithm.AggregatorArrayProvider provider = new BaseTopNAlgorithm.AggregatorArrayProvider(
+        (DimensionSelector) params.getDimSelector(),
+        query,
+        params.getCardinality(),
+        capabilities
+    );
+
+    // Unlike regular topN we cannot rely on ordering to optimize.
+    // Optimization possibly requires a reverse lookup from value to ID, which is
+    // not possible when applying an extraction function
+    return provider.build();
+  }
+
+  @Override
+  public void dimExtractionScanAndAggregate(Object selector, Aggregator[][] rowSelector, Map<Comparable, Aggregator[]> aggregatesStore, Cursor cursor, TopNQuery query)
+  {
+    final DimensionSelector dimSelector = (DimensionSelector) selector;
+    final IndexedInts dimValues = dimSelector.getRow();
+
+    for (int i = 0; i < dimValues.size(); ++i) {
+      final int dimIndex = dimValues.get(i);
+      Aggregator[] theAggregators = rowSelector[dimIndex];
+      if (theAggregators == null) {
+        final String key = dimSelector.lookupName(dimIndex);
+        theAggregators = aggregatesStore.get(key);
+        if (theAggregators == null) {
+          theAggregators = BaseTopNAlgorithm.makeAggregators(cursor, query.getAggregatorSpecs());
+          aggregatesStore.put(key, theAggregators);
+        }
+        rowSelector[dimIndex] = theAggregators;
+      }
+
+      for (Aggregator aggregator : theAggregators) {
+        aggregator.aggregate();
+      }
+    }
+  }
+
+
+  @Override
+  public void addRowValuesToSelectResult(String outputName, Object dimSelector, Map<String, Object> theEvent)
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+    if (selector == null) {
+      theEvent.put(outputName, null);
+    } else {
+      final IndexedInts vals = selector.getRow();
+
+      if (vals.size() == 1) {
+        final String dimVal = selector.lookupName(vals.get(0));
+        theEvent.put(outputName, dimVal);
+      } else {
+        List<String> dimVals = Lists.newArrayList();
+        for (int i = 0; i < vals.size(); ++i) {
+          dimVals.add(selector.lookupName(vals.get(i)));
+        }
+        theEvent.put(outputName, dimVals);
+      }
+    }
+  }
+
+  @Override
+  public void updateSearchResultSet(String outputName, Object dimSelector, SearchQuerySpec searchQuerySpec, TreeMap<SearchHit, MutableInt> set, int limit)
+  {
+    final DimensionSelector selector = (DimensionSelector) dimSelector;
+
+    if (selector != null) {
+      final IndexedInts vals = selector.getRow();
+      for (int i = 0; i < vals.size(); ++i) {
+        final String dimVal = selector.lookupName(vals.get(i));
+        if (searchQuerySpec.accept(dimVal)) {
+          MutableInt counter = new MutableInt(1);
```

Comment:
This logic with `MutableInt` replacement and allocation is ugly, but unfortunately the only way to increment the value making only one map search in Java 7. Maybe can use `Object2IntRBTreeMap`?